### PR TITLE
[WIP] Added cue type definition to typescript types

### PR DESCRIPTION
**Pull request by mehalter**
Started adding ability to convert Cue to TS types to maintain a single source of truth for the typing of contracts, blueprints, etc.

### CODE REVIEW

File: scripts/build-types.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	const cueresult = await cue('export', [input], { '--out': 'openapi' });
+	if (cueresult.code !== 0) {
+		console.error(cueresult.stderr);
+		throw new Error('Error exporting cue as openapi');
+	}
+	const openapi = JSON.parse(cueresult.stdout);
+	return openapiTS(openapi);
+};
+
+(async () => {
```
Comment (codewithcheese) [inline scripts/build-types.ts:16]: Declared function as `async` but never used `await`. I suggest rewriting the function body to use try/await/catch

File: scripts/build-types.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import cue from 'cuelang-js';
+import { default as openapiTS } from 'openapi-typescript';
```
Comment (codewithcheese) [inline scripts/build-types.ts:2]: `import` without `{}` is default. `import openapiTS from 'openapi-typescript';` I would call it `openApiToTS`

File: scripts/build-types.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import { promises as fs } from 'fs';
+
+const cueToTSTypes = async (input: string) => {
+	const cueresult = await cue('export', [input], { '--out': 'openapi' });
```
Comment (codewithcheese) [inline scripts/build-types.ts:7]: Multiple words should use camelCase. I would call it `exportResult`

File: scripts/build-types.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+	if (cueresult.code !== 0) {
+		console.error(cueresult.stderr);
+		throw new Error('Error exporting cue as openapi');
```
Comment (codewithcheese) [inline scripts/build-types.ts:10]: Thrown error will result in call to `console.error`, above `console.error` seems redundant. Can include the `stderr` in the `Error` message. 

```
Error(`Failed to export cue to openapi: ${cueresult.stderr}`);
```

File: package.json [hunk 5]
```diff
@@ -9,12 +9,13 @@
     "url": "git@github.com:resin-io-modules/contrato.git"
   },
   "scripts": {
+    "buildtypes": "npx ts-node@^8.0.0 ./scripts/build-types.ts",
```
Comment (codewithcheese) [inline package.json:12]: Is `@^8.0.0` necessary?

**Review by codewithcheese — COMMENTED** [review]
When I run `npm run buildtypes` I get the following errors:

```
⨯ Unable to compile TypeScript:
scripts/build-types.ts:4:32 - error TS2307: Cannot find module 'fs' or its corresponding type declarations.

4 import { promises as fs } from 'fs';
                                 ~~~~
scripts/build-types.ts:16:2 - error TS2705: An async function or method in ES5/ES3 requires the 'Promise' constructor.  Make sure you have a declaration for the 'Promise' constructor or include 'ES2015' in your `--lib` option.

16 (async () => {
```

Do you get the same errors?

**Comment by codewithcheese** [comment]
This change will probably cause type errors for users of `Contract` since their `ContractObject` will not match the definition. This will probably have to be a `major` release. The tests might also have type errors if they were typescript. We should probably convert this whole project to typescript since we're attempting to be more formal about typings.

File: scripts/build-types.ts (hunk 1 above)
Comment (mehalter) [reply scripts/build-types.ts]: removed anonymous async function call in latest commit

File: scripts/build-types.ts (hunk 2 above)
Comment (mehalter) [reply scripts/build-types.ts]: Fixed in latest commit

File: scripts/build-types.ts (hunk 3 above)
Comment (mehalter) [reply scripts/build-types.ts]: Fixed in latest commit

File: scripts/build-types.ts (hunk 4 above)
Comment (mehalter) [reply scripts/build-types.ts]: Fixed in latest commit

File: package.json (hunk 5 above)
Comment (mehalter) [reply package.json]: I was thinking to protect against breaking changes, but we aren't using any flags, so it should be fine to drop it

**Comment by mehalter** [comment]
Included in #38

**Pull request closed without merging**